nbinteract/figure.py

```diff
@@ -0,0 +1,259 @@
+        self.functions = functions
+        self.marks = marks.values()
+        self.figure = plotting._create_fig_with_options(options=options)
+        self.plot = plotting._create_fig_with_options(options=options)
+        self.create_plot()
+
+    def create_plot(self):
+        """
+        Generates the appropriate plot markers for each marker in the marks
+        dictionary and links them with the specified widgets.
+        """
+        for mark_features_dict in self.marks:
+            mark_name = mark_features_dict['marker']
+            mark_fn = Plot.NAME_TO_FN[mark_name]
+            # add in self.figure to options
+            mark_features_dict['options'] = tz.assoc(
+                mark_features_dict['options'], '_fig', self.figure
+            )
+            # removed mark from dictionary because it's not a param of
+            # the plotting function
+            no_marks_dict = self.minus_key('marker', mark_features_dict)
+            wrapped_fns = mark_fn(**no_marks_dict)
+            if mark_features_dict.get('x_fn'):
+                self.link_interactive(
+                    mark_features_dict['x_fn'], wrapped_fns[0]
+                )
+            if mark_features_dict.get('y_fn'):
+                self.link_interactive(
+                    mark_features_dict['y_fn'], wrapped_fns[1]
+                )
+        return self
+
+    def link_interactive(self, original_fn, wrapped_fn):
+        """
+        Creates a trailet link that maps the wrapper function to the
+        appropriate widget as specified in the widgets dictionary. The widget
+        specified by the API for the original_function will now be linekd
+        with the wrapped function which allows dynamic updating.
+        """
+        if callable(original_fn):
+            req_arguments = util.get_required_args(original_fn)
+            req_widget_names = self.functions.get(original_fn)
+            argument_dict = dict(
+                zip(
+                    req_arguments,
+                    [self.widgets.get(name) for name in req_widget_names]
+                )
+            )
+            interactive_obj = widgets.interactive(wrapped_fn, **argument_dict)
+            initial_values = dict(
+                zip(
+                    req_arguments, [
+                        self.widgets.get(name).get_interact_value()
+                        for name in req_widget_names
+                    ]
+                )
+            )
+            wrapped_fn(**initial_values)
+        else:
+            wrapped_fn()
+
+    def _ipython_display_(self):
+        """
+        Called when a Plot is returned on the last line of a Jupyter cell to
+        automagically display the widgets and Figure.
+        """
+        display(widgets.VBox(list(self.widgets.values()) + [self.figure]))
+
+    def minus_key(self, key, dictionary):
+        """
+        Creates copy of a dictionary with the specified key/value pair removed.
+        """
```

Comment:
You can replace this function with http://toolz.readthedocs.io/en/latest/api.html#toolz.dicttoolz.dissoc .

Reply:
done!